application.py

```diff
@@ -1,13 +1,478 @@
+    auth = (AWS_COGNITO_CLIENT_ID, AWS_COGNITO_CLIENT_SECRET)
+
+    response = requests.post(token_url, data=payload, headers=headers, auth=auth)
+    response_json = response.json()
+    if response.status_code != 200:
+        if response_json.get("error") == "invalid_grant":
+            logger.error("Invalid authorization code")
+            return {"error": "Invalid authorization code"}
+        logger.error(f"Error: {response.status_code}, {response.text}")
+        return {"error": response_json}
+    else:
+        # Setting the session tokens to be used in the session later in home().abort
+        return response.json()
+
+
+def get_email():
+    try:
+        headers = {"Authorization": f"{session['id_token']}"}
+        user_request = requests.get(
+            f"{API_URL}/api/user",
+            headers=headers,
+        )
+        if user_request.status_code != 200:
+            return False
+        else:
+            session["email"] = user_request.json()["email"]
+            return True
+    except Exception as error:
+        logger.error(f"{error.__class__.__name__}: {error}")
+        return False
+
+
+@app.route("/dashboard", methods=["GET"])
+def dashboard():
+    headers = {"Authorization": f"{session['id_token']}"}
+    projects = requests.get(
+        f"{API_URL}/api/projects",
+        headers=headers,
+    ).json()
+    try:
+        if len(projects) == 0:
+            flash("You have no projects. Get started by clicking 'Create a Project'.")
+        return render_template(
+            "dashboard.html", email=session["email"], projects=projects
+        )
+    except Exception as error:
+        logger.error(error)
+        flash("Please re-login to authenticate your account.")
+        return redirect(url_for("home"))
+
+
+@app.route("/project/<project_name>", methods=["GET"])
+def view_project(project_name):
+    headers = {"Authorization": f"{session['id_token']}"}
+    projects = requests.get(
+        f"{API_URL}/api/projects/{project_name}",
+        headers=headers,
+    ).json()
+    # projects either returnes {'description': 'Project not found', 'message': None} or a project in dict form.
+    print(projects)
+    try:
+        if projects["message"] and projects["message"] is None:
+            flash("Project not found. Please try again.")
+            return redirect(url_for("dashboard"))
+    except Exception:
+        return render_template("view_project.html", project=projects)
+
+
+@app.route("/survey", methods=["GET", "POST"])
+def survey():
+    # ONLY GET AND POST ARE ALLOWED HENCE NO NEED FOR SECOND IF STATEMENT
+    # IF METHOD IS 'GET' THEN THE SURVEY IS RENDERED
+    if request.method == "GET":
+        return render_template("survey.html")
+    # IF METHOD IS 'NOT GET' THEN THE POST PROCESS BEGINS
+    headers = {
+        "Authorization": f"{session['id_token']}",
+        "content-type": "application/json",
+    }
+
+    user = json.loads(request.form["user"])
```

Comment:
Can we create a map_form_data() function for this. It'll improve readability.

Reply:
Created 👍